### PR TITLE
build tarfiles for docker images in a separate process

### PR DESCRIPTION
**Pull request by jonmoter**
Creates a tarfile in a forked process

/cc @zendesk/samson

### Tasks
 - [ ] :+1: from team

### References
 - Jira link: https://zendesk.atlassian.net/browse/SAMSON-303

### Risks
- Level: Med

### CODE REVIEW

File: app/models/docker_builder_service.rb [hunk 1]
```diff
@@ -9,17 +9,43 @@ class DockerBuilderService
...
         output.write_docker_chunk(chunk)
       end
     output.puts('### Docker build complete')
 
     docker_image
   end
 
+  def self.create_docker_tarfile(dir)
+    dir += '/' unless dir.end_with?('/')
+    tempfile_name = Dir::Tmpname.create('out') {}
```
Comment (grosser) [inline app/models/docker_builder_service.rb]: normally the dir is cleaned up at the end of the block ?

Comment (jonmoter) [inline app/models/docker_builder_service.rb]: Good point.  Will clean up.

File: app/models/docker_builder_service.rb [hunk 2]
```diff
@@ -9,17 +9,43 @@ class DockerBuilderService
...
+    # It's possible for the puma process to seize up doing all those syscalls,
+    # especially if the disk is running slow. So we create the tarfile in a
+    # separate process to avoid that.
+    tar_proc = -> do
+      File.open(tempfile_name, 'wb+') do |tempfile|
+        Docker::Util.create_relative_dir_tar(dir, tempfile)
+      end
+    end
+
+    if ENV['RAILS_ENV'] == 'test'
```
Comment (grosser) [inline app/models/docker_builder_service.rb]: `Rails.env.test?`

**Comment by grosser** [comment]
👍

**Comment by sandlerr** [comment]
The [fork doc](https://ruby-doc.org/core-2.1.2/Process.html#method-c-fork) suggests using `spawn` instead, that seems like it might be worth considering since it allows setting rlimits.

File: app/models/docker_builder_service.rb [hunk 3]
```diff
@@ -9,15 +9,46 @@ class DockerBuilderService
...
 
     docker_image
+  ensure
+    if tarfile
+      tarfile.close
+      FileUtils.rm(tarfile.path, force: true)
+    end
+  end
+
+  def self.create_docker_tarfile(dir)
```
Comment (grosser) [inline app/models/docker_builder_service.rb]: `private_class_method def self.`

**Comment by sandlerr** [comment]
🎉 thanks Jon!

**Pull request merged**